phylo/acceleration.go

```diff
@@ -0,0 +1,344 @@
+	sub.VhumHca = (sub.DhumChimp + sub.DhumHga - sub.DchimpHga) / 2.0
+	sub.VchimpHca = (sub.DhumChimp + sub.DchimpHga - sub.DhumHga) / 2.0
+	sub.VhcaHga = (sub.DhumHga + sub.DchimpHga - sub.DhumChimp) / 2.0
+	if allowNegative {
+		answer.BhumHca = sub.VhumHca
+		answer.BchimpHca = sub.VchimpHca
+		answer.BhcaHga = sub.VhcaHga
+		return
+	}
+	if sub.VhumHca < 0 && sub.VchimpHca < 0 && sub.VhcaHga < 0 {
+		if verbose {
+			log.Printf("WARNING: All branches are negative.") //TODO: Should this error out?
+		}
+		sub.VhumHca = 0
+		sub.VchimpHca = 0
+		sub.VhcaHga = 0
+	} else if sub.VhumHca < 0 {
+		sub.VhumHca = 0
+		if sub.VchimpHca < 0 {
+			sub.VchimpHca = 0
+			sub.VhcaHga = nonNegativeApproximation(sub.DhumHga, sub.DchimpHga, sub.VhumHca, sub.VchimpHca, sub.WhumHga, sub.WchimpHga)
+		} else if sub.VhcaHga < 0 {
+			sub.VhcaHga = 0
+			sub.VchimpHca = nonNegativeApproximation(sub.DhumChimp, sub.DchimpHga, sub.VhumHca, sub.VhcaHga, sub.WhumChimp, sub.WchimpHga)
+		} else {
+			sub.VhcaHga = nonNegativeApproximation(sub.DhumHga, sub.DchimpHga, sub.VhumHca, sub.VchimpHca, sub.WhumHga, sub.WchimpHga)
+			sub.VchimpHca = nonNegativeApproximation(sub.DhumChimp, sub.DchimpHga, sub.VhumHca, sub.VhcaHga, sub.WhumChimp, sub.WchimpHga)
+		}
+	} else if sub.VchimpHca < 0 {
+		sub.VchimpHca = 0
+		if sub.VhcaHga < 0 {
+			sub.VhcaHga = 0
+			sub.VhumHca = nonNegativeApproximation(sub.DhumHga, sub.DhumChimp, sub.VhcaHga, sub.VchimpHca, sub.WhumHga, sub.WhumChimp)
+		} else {
+			sub.VhumHca = nonNegativeApproximation(sub.DhumHga, sub.DhumChimp, sub.VhcaHga, sub.VchimpHca, sub.WhumHga, sub.WhumChimp)
+			sub.VhcaHga = nonNegativeApproximation(sub.DhumHga, sub.DchimpHga, sub.VhumHca, sub.VchimpHca, sub.WhumHga, sub.WchimpHga)
+		}
+	} else if sub.VhcaHga < 0 {
+		sub.VhcaHga = 0
+		sub.VhumHca = nonNegativeApproximation(sub.DhumHga, sub.DhumChimp, sub.VhcaHga, sub.VchimpHca, sub.WhumHga, sub.WhumChimp)
+		sub.VchimpHca = nonNegativeApproximation(sub.DhumChimp, sub.DchimpHga, sub.VhumHca, sub.VhcaHga, sub.WhumChimp, sub.WchimpHga)
+	}
+	answer.BhumHca = sub.VhumHca
+	answer.BchimpHca = sub.VchimpHca
+	answer.BhcaHga = sub.VhcaHga
+}
+
+//optimizeSubtreeRight is a helper function of BranchLengthsAlternatingLeastSquares and computes the optimal branch lengths for the right subtree at a particular iteration.
+func optimizeSubtreeRight(sub *AccelSubTreeRight, allowNegative bool, verbose bool, answer *AccelBranchLengths) {
+	sub.VhcaHga = (sub.DhcaGor + sub.DhcaOrang - sub.DgorOrang) / 2.0
+	sub.VhgaGor = (sub.DhcaGor + sub.DgorOrang - sub.DhcaOrang) / 2.0
+	sub.VhgaOrang = (sub.DhcaOrang + sub.DgorOrang - sub.DhcaGor) / 2.0
+	if allowNegative {
+		answer.BhcaHga = sub.VhcaHga
+		answer.BhgaGor = sub.VhgaGor
+		answer.BhgaOrang = sub.VhgaOrang
+		return
+	}
+
+	if sub.VhcaHga < 0 && sub.VhgaGor < 0 && sub.VhgaOrang < 0 {
+		if verbose {
+			log.Printf("WARNING: All branches are negative.") //TODO: Should this error out?
```

Comment:
For the places where all branches are negative, are the negative results meaningful? Do they just need some kind of conversion to be the correct positive numbers again? Have you run into this before?